### DISCUSSION TITLE
Business rules - stop triggering

### DISCUSSION THREAD

**Discussion by BlagoCuljak** [MarimerLLC/csla#1716] [Questions] [answered]
Hello everyone!

Since the upgrade to CSLA 5x, we are occasionaly experiencing a very strange (and dangerous...) rules behavior. Namely, without any apparent reason all of the business rules stop triggering completely. I am not talking about a single busines object here, but – to be clear – once this happens all of the objects (unrelated to the one where the problem started) are affected. Only after shutting down the entire application and starting it again normal behavior is restored. It happens very rarely and we cannot reproduce the issue, but I kind of hope someone here has at least some vague idea what could possibly cause such unexpected and undesirable effects.

**Comment by rockfordlhotka** [comment, marked as answer]
What kind of app?

**Reply by BlagoCuljak** [reply]
WPF, to make matter worst, the DP is running as it should, and Object saves into database, just the I can save a 2 milion value, and Rule is 100 tops.

**Reply by rockfordlhotka** [reply]
That's weird for sure! Is the device under memory pressure? Or some sort of threading pressure from the app?

**Reply by BlagoCuljak** [reply]
Do you mean WPF app or AppServer?

**Reply by JacoJordaan** [reply]
I had a similar issue with WPF where rules stop running. 

This happened when I had async rules. My short term solution was just to make sure I do not have any async rules.

 I could never pinpoint the exact cause, but it is threading related.

**Reply by JacoJordaan** [reply]
Just to clarify, this was an issue for me before version 5.x. So, it may not be related.

**Reply by rockfordlhotka** [reply]
@BlagoCuljak I mean the WPF app - I assume that's where the rules are failing?

**Reply by BlagoCuljak** [reply]
Well, I will try to see memory footprint, next time this happens.

@JacoJordaan have you upgraded to 5.x? Has this occurred again then?

**Reply by JacoJordaan** [reply]
I have upgraded to 5.2, but I have not tested it again with async rules.

If I remember correctly the rules stopped when I interacted with the WPF UI client while a async rule was running.

**Reply by BlagoCuljak** [reply]
Happened once more this morning. The entire app was 312MB in Task Manager, 8GB available. 

We have some async rules, and we will try to change them. 

We didn't had this behavior with 4.6, WPF was more or less same code base as is with 5.2.

**Reply by rockfordlhotka** [reply]
Please keep me updated @BlagoCuljak - clearly there's some issue, and if we (you?) can narrow it down I'll enthusiastically help address it!!

**Reply by rockfordlhotka** [reply]
@BlagoCuljak I was looking through code history trying to find anything obvious. Nothing jumps out at me.

However, I am wondering if maybe `ISuppressRuleChecking` are in use? And maybe an exception happens where these are used, such that rule checking never gets turned back on? I'm guessing the answer is no, but it is just an idea.

**Reply by BlagoCuljak** [reply]
Hi, I'm out of office today, but my co-worked said that he has been testing ISuppressRuleChecking, and maybe its leftover in code.

We will double check,and see our findings.

**Reply by BlagoCuljak** [reply]
Well it's not ISuppressRuleChecking, we don't have it in code anywhere...

**Reply by BlagoCuljak** [reply]
Well, just to let you know that we are still having issues, removed the async rules, upgraded WPF to 4.7.2, no logs or errors when this happens. We have reproduces this twice in VS, but no error is thrown. The app just doesn't compute as it should.

We are desperate, and we will include CSLA as project with code, not as NuGet, but we are not sure if we will pinpoint the issue in that way.

**Reply by rockfordlhotka** [reply]
Referencing CSLA directly may help. You might put console.writeline (or other log) statements into the code that executes the rules to help see what's happening? Since this is not repeatable, using break points in the debugger probably won't be efficient.

It isn't clear to me how I can help until the problem is narrowed down, but I'll help when possible.

**Reply by BlagoCuljak** [reply]
Well, we found culprit.. One Async rule had error, and although it had been in try/catch, on that one rule, we didn't logged error. Talk about bad luck.

It's all good once again.

**Reply by rockfordlhotka** [reply]
@BlagoCuljak do you think there is anything CSLA could have done to help make this more obvious?

**Reply by BlagoCuljak** [reply]
I don't think so. This was essential mistake in out code and arguments that Rule was working with.

Bad code review, no logging in catch, and so on.

Although same rule did work in 4.7, we are glad that this issue was brought to us when we updated to 5.x. 

Good job Rocky (as always).
